accepted/system-range/system-range.md

```diff
@@ -0,0 +1,200 @@
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting negative ranges or element types
+  other than `Int32`).
+
+### Non-Goals
+
+* Supporting other numeric types or even offer a generic range type
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+
+## Design
+
+### Proposed API
+
+```csharp
+namespace System {
+    public struct Range : IEnumerable, IEnumerable<int>, IEquatable<Range> {
+        public static Range Empty { get; }
+        public static Range All { get; }
+        public static Range Between(int start, int end);
+        public static Range From(int start);
+        public static Range To(int end);
+        public Range(int index, int length);
+        public int Index { get; }
+        public int Length { get; }
+        public bool IsClosed { get; }
+        public Range Close(Range range);
+        public override string ToString();
+        public override int GetHashCode();
+        public bool Equals(Range other);
+        public override bool Equals(object other);
+        public static bool operator==(Range left, Range right);
+        public static bool operator!=(Range left, Range right);
+    }
+}
+```
+
+### Companion APIs
+
+```csharp
+namespace System {
+    public partial class String {
+        public string this[Range range] { get; }
+    }
+    public partial class T[] {
+        public T[] this[Range range] { get; }
+    }
+    public partial struct ArraySegment<T> {
+        public ArraySegment<T> this[Range range] { get; }
+    }      
+    public partial struct Span<T> {
+        public Span<T> this[Range range] { get; }
+    }
+}
+```
+
+## Q & A
+
+### Open Design Points
+
+* Right now, we cannot retreive `Index` or `Length` if the range is open on any
+  side (it throws `NotSupportedException`).
+    - We could expose `Start: int?` and `End: int?` and then add a factory
+      method `Between(int? start, int? end)`
+
+* We should check whether we can handle negative numbers for `Length` so that we
+  can support addressing relative to the end. Check Python to see what they did
+  there.
+
+* Should `Range` support containment checks?
+    - Positions and/or ranges?
+    - Question is how to define semantics for open ended ranges.
+
+* Should `Range` support combining/intersecting open ended ranges?
+    - Be careful though. For open ended slices this might produce invalid
+      values. For instance, you might think combining `3..` and `..12` would
+      produce `3..12` but this would mean indexing into an array of size 5 would
+      fail, while `3..` would have worked.
+    - We could define the operation as `Close(Range)` though.
```

Comment:
I would personally always assume that we close the range over the collection first before attempting any range-based indexing. Just taking pagination as an example: `items[new Range(page * pageSize, length: pageSize)]`. I would totally expect this to return less than `pageSize` items on the last page when there aren’t enough items, instead of failing.

Reply:
> I would personally always assume that we close the range over the collection first before attempting any range-based indexing.

Sort of. I expect the indexer of array to take care of it. But the indexer will throw an argument out of range exception if the resulting range is invalid. And for `3..` it would close it to `3..5` and succeed while `3..12` would fail. Hence, thinking of intersecting is logically wrong.